### PR TITLE
Paganition Implementation

### PR DESCRIPTION
**Pull request by yxie66**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have already squashed the commits and make the commit message conform to the project standard.
- [ ] I have already marked the commit with "BREAKING CHANGE" or "Fixes #" if needed.
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: dataprep/data_connector/connector.py [hunk 1]
```diff
@@ -76,6 +77,7 @@ def __init__(
             ensure_config(config_path)
             path = config_directory() / config_path
             self.impdb = ImplicitDatabase(path)
+        # print(self.impdb.tables)
```
Comment (dovahcrow) [inline dataprep/data_connector/connector.py]: Please remove unnecessary comments

File: dataprep/data_connector/implicit_database.py [hunk 2]
```diff
@@ -67,8 +67,9 @@ def __init__(self, name: str, config: Dict[str, Any]) -> None:
         )  # This will throw errors if validate failed
         self.name = name
         self.config = config
-
         request_def = config["request"]
+        if config["request"]["paganition"]["type"] == "limit":
```
Comment (dovahcrow) [inline dataprep/data_connector/implicit_database.py]: I didn't see any usage of these two lines

File: dataprep/data_connector/implicit_database.py [hunk 3]
```diff
@@ -87,6 +88,7 @@ def __init__(self, name: str, config: Dict[str, Any]) -> None:
 
         for key in ["headers", "params", "cookies"]:
             if key in request_def:
+
```
Comment (dovahcrow) [inline dataprep/data_connector/implicit_database.py]: Do not have unnecessary modifications.

File: dataprep/data_connector/implicit_database.py [hunk 4]
```diff
@@ -230,10 +232,9 @@ def __init__(self, config_path: Union[str, Path]) -> None:
             if table_config_path.suffix != ".json":
                 # ifnote non json file
                 continue
-
             with open(table_config_path) as f:
                 table_config = jload(f)
-
+                self.config = table_config
```
Comment (dovahcrow) [inline dataprep/data_connector/implicit_database.py]: What's this for? If you need to create a new instance variable, please also define it in the class header for easy reading.

File: dataprep/data_connector/schema.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
         "request",
         "response"
     ],
-    "additionalProperties": false,
+    "additionalProperties": true,
```
Comment (dovahcrow) [inline dataprep/data_connector/schema.json]: If you'd like to add new fields to the config file, please write it explicitly. `additionalProperties: true` is not a good practice.

File: test.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
```
Comment (dovahcrow) [inline test.py]: If it is a test file, please put it under the tests folder with certain filesystem structures like how EDA did.

**Comment by yxie66** [comment]
> Please don't commit test.py as it contains sensitive API keys.

got it, thanks

File: dataprep/data_connector/implicit_database.py (hunk 2 above)
Comment (yxie66) [reply dataprep/data_connector/implicit_database.py]: removed.

File: dataprep/data_connector/schema.json [hunk 7]
```diff
@@ -61,6 +61,9 @@
                 "params": {
                     "$ref": "#/definitions/fields"
                 },
+                "paganition": {
```
Comment (dovahcrow) [inline dataprep/data_connector/schema.json:64]: Pagination is not a field. Please define the fields in place.

File: dataprep/data_connector/schema.json [hunk 8]
```diff
@@ -145,6 +148,9 @@
                     {
                         "type": "boolean"
                     },
+                    {
```
Comment (dovahcrow) [inline dataprep/data_connector/schema.json:151]: The fields definition is the definition of a bunch of fields. Currently it means the fields def can be:
```
{
    "fieldA": "template", // a string template
    "fieldB": false // not required
    "fieldC": {
    // a complete field definition
    }
}
```
Please do not change the fields definition.

File: dataprep/data_connector/schema.json [hunk 9]
```diff
@@ -145,6 +148,9 @@
...
+                        "type": "integer"
+                    },
                     {
                         "type": "object",
                         "required": [
```
Comment (dovahcrow) [inline dataprep/data_connector/schema.json:156]: The table configs should be a separate PR to DataConnectorConfigs repo. 
Also, please squash the PR and rebase to the `develop` branch.

Thanks!

File: dataprep/data_connector/connector.py [hunk 10]
```diff
@@ -136,6 +162,98 @@ def _fetch(
...
+        table : str
+            The table name.
+        _auth : Optional[Dict[str, Any]] = None
+            The parameters for authentication. Usually the authentication parameters
+            should be defined when instantiating the Connector. In case some tables have different
+            authentication options, a different authentication parameter can be defined here.
+            This parameter will override the one from Connector if passed.
+        _count: Optional[int] = None,
+            count of returned records.
+        **where: Dict[str, Any]
```
Comment (dovahcrow) [inline dataprep/data_connector/connector.py:186]: The documentation is wrong

File: dataprep/data_connector/connector.py [hunk 11]
```diff
@@ -136,6 +162,98 @@ def _fetch(
...
+            The additional parameters required for the query.
+
+        Returns
```
Comment (dovahcrow) [inline dataprep/data_connector/connector.py:189]: No need for the Returns statement. Check our wiki for docstring best practice https://github.com/sfu-db/dataprep/wiki/Docstring-Best-Practice

File: dataprep/data_connector/implicit_database.py [hunk 12]
```diff
@@ -54,6 +54,7 @@ class ImplicitTable:  # pylint: disable=too-many-instance-attributes
     body_ctype: str
     body: Optional[Fields] = None
     cookies: Optional[Fields] = None
+    pag_params: Dict[str, Any]
```
Comment (dovahcrow) [inline dataprep/data_connector/implicit_database.py:57]: Can you flesh out the Dict into a deterministic class? It is so easy to make mistakes with this loosely defined dict.

File: dataprep/data_connector/schema.json [hunk 13]
```diff
@@ -61,6 +61,19 @@
                 "params": {
                     "$ref": "#/definitions/fields"
                 },
+                "pagination": {
+                    "$id": "#/properties/request/properties/pagination",
+                    "type":"object",
+                    "additionalProperties": {
```
Comment (dovahcrow) [inline dataprep/data_connector/schema.json:67]: Can you make the keys of pagination deterministic and forbid additional properties?

File: dataprep/data_connector/implicit_database.py [hunk 14]
```diff
@@ -54,6 +54,7 @@ class ImplicitTable:  # pylint: disable=too-many-instance-attributes
     body_ctype: str
     body: Optional[Fields] = None
     cookies: Optional[Fields] = None
+    pag_params: Dict[str, str]
```
Comment (dovahcrow) [inline dataprep/data_connector/implicit_database.py:57]: Because pagination parameters are already clearly defined and deterministic, please change it to `pag_params: Optional[Pagination]` and define your own `class Pagination` to make the code clear.

File: dataprep/data_connector/schema.json [hunk 15]
```diff
@@ -61,6 +61,15 @@
                 "params": {
                     "$ref": "#/definitions/fields"
                 },
+                "pagination": {
+                    "$id": "#/properties/request/properties/pagination",
```
Comment (dovahcrow) [inline dataprep/data_connector/schema.json:65]: Please format this json file.
Please add `required` field, and `"additionalProperties": false` to make this robust.

**Comment by dovahcrow** [comment]
Thanks @yxie66! I think now this PR can be merged with several format changes.

**Comment by dovahcrow** [comment]
Hi @yxie66, I greatly refactored your code. The original code has some edge cases not considered and a few pieces of redundant code. Please take a look at my commit (bce9ad8) to understand the changes I made.

**Comment by dovahcrow** [comment]
Waiting on sfu-db/DataConnectorConfigs#12.
